tools/syz-crush/crush.go

```diff
@@ -24,107 +28,252 @@ import (
+					runType)
+				if atomic.LoadUint32(&shutdown) != 0 || !*flagInfinite {
+					atomic.AddUint32(&runningWorkers, ^uint32(0))
+
+					// If this is the last worker then we can close the channel
+					if atomic.LoadUint32(&runningWorkers) == 0 {
+						log.Logf(0, "vm-%v: closing channel", index)
+						close(runDone)
+					}
 					break
+				} else {
+					log.Logf(0, "vm-%v: restarting", index)
 				}
 			}
-		}()
+			log.Logf(0, "vm-%v: done", index)
+		}(i)
 	}
 
+	log.Logf(0, "restart/timeout set to: %v", *flagRestartTime)
 	shutdownC := make(chan struct{})
 	osutil.HandleInterrupts(shutdownC)
 	go func() {
 		<-shutdownC
-		wg.Done()
 		atomic.StoreUint32(&shutdown, 1)
 	}()
-	wg.Wait()
+
+	var count int
+	var crashes int
+	for res := range runDone {
+		count++
+		crashes += storeCrash(res, cfg, baseName)
+		log.Logf(0, "instances executed: %v", count)
+	}
+
+	log.Logf(0, "all done. reproduced %v crashes. reproduce rate %.2f%%", crashes, float64(crashes)/float64(count)*100.0)
 }
 
-func runInstance(cfg *mgrconfig.Config, reporter report.Reporter, vmPool *vm.Pool, index int) {
+func storeCrash(res *CrashReport, cfg *mgrconfig.Config, baseName string) int {
+	log.Logf(0, "storing results...")
+	if res == nil || res.Report == nil {
+		log.Logf(0, "nothing to store")
+		return 0
+	}
+
+	log.Logf(0, "loop: instance %v finished, crash=%v", res.vmIndex, res.Report.Title)
+
+	crashdir := filepath.Join(cfg.Workdir, "crashes")
+	osutil.MkdirAll(crashdir)
+
+	sig := hash.Hash([]byte(res.Report.Title))
+	id := sig.String()
+	dir := filepath.Join(crashdir, id)
+	log.Logf(0, "vm-%v: crashed: %v, saving to %v", res.vmIndex, res.Report.Title, dir)
+
+	osutil.MkdirAll(dir)
+	if err := osutil.WriteFile(filepath.Join(dir, "description"), []byte(res.Report.Title+"\n")); err != nil {
+		log.Logf(0, "failed to write crash: %v", err)
+	}
+	// Save up to 100 reports. If we already have 100, overwrite the oldest one.
+	// Newer reports are generally more useful. Overwriting is also needed
+	// to be able to understand if a particular bug still happens or already fixed.
+	oldestI := 0
+	var oldestTime time.Time
+	for i := 0; i < 100; i++ {
+		info, err := os.Stat(filepath.Join(dir, fmt.Sprintf("log%v", i)))
+		if err != nil {
+			oldestI = i
+			break
+		}
+		if oldestTime.IsZero() || info.ModTime().Before(oldestTime) {
+			oldestI = i
+			oldestTime = info.ModTime()
+		}
+	}
+	osutil.WriteFile(filepath.Join(dir, fmt.Sprintf("log%v", oldestI)), res.Report.Output)
+	if len(cfg.Tag) > 0 {
+		osutil.WriteFile(filepath.Join(dir, fmt.Sprintf("tag%v", oldestI)), []byte(cfg.Tag))
+	}
+	if len(res.Report.Report) > 0 {
+		osutil.WriteFile(filepath.Join(dir, fmt.Sprintf("report%v", oldestI)), res.Report.Report)
+	}
+
+	reproducedWithdir := filepath.Join(dir, "reproduced_with")
+	osutil.MkdirAll(reproducedWithdir)
+	if err := osutil.WriteFile(filepath.Join(reproducedWithdir, baseName), []byte(baseName+"\n")); err != nil {
```

Comment:
With a small change, we can make tag contain the baseName for each crash. Then we don't diverge from the manager format at all.